include/envoy/common/account.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+
+class Charge;
+
+/*
+ * An interface for accounting the usage of a resource.
+ * The "resource" tracked here is implicit, if in the future we want to track multiple
+ * resources, we should expand the interface to make it explicit.
+ *
+ * Currently this is only used by L7 streams to track the amount of memory used
+ * by the stream.
+ *
+ */
+class Account {
+public:
+  virtual ~Account() = default;
+
+  /**
+   * Returns the outstanding balance.
+   */
+  virtual uint64_t balance() const PURE;
```

Comment:
Should balance() be part of the public interface?  It seems to me that the consumers of the charge/credit methods and balance are distinct.  The expected caller of the balance method is the HTTP stream object who needs to decide if memory usage is within allowed bounds.  Callers of charge/credit are the buffers used by the stream.

Reply:
Given the use of composition (i.e. stream is among the owner of `BufferMemoryAccount`) it seem to make sense that `balance()` is public.)

Originally I tried to make `Stream` themselves implement account, but that was not the right direction.

Reply:
It makes sense to me for balance() to be a method of BufferMemoryAccount.  It doesn't need to be a method of Account.

Reply:
Can we remove balance() from the API until we have proof that it is needed?

Reply:
Done.